## Gather

In [1]:
import pandas as pd

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [3]:
patients.head(6)

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   
5           6         male     Rafael    Costa         1140 Willis Avenue   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   
5      Daytona Beach     Florida   32114.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1  
5    386-334-5237RafaelCardosoCosta@gustr.com  8/31/1931   183.9      70  26.4

In [4]:
treatments.head(10)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   
5    jasmine       sykes          -  42u - 44u         7.56       7.18   
6     sophia      haugen  37u - 42u          -         7.65       7.27   
7      eddie      archer  31u - 38u          -         7.89       7.55   
8      saber      ménard          -  54u - 54u         8.08       7.70   
9       asia     woźniak  30u - 36u          -         7.76       7.37   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32  
5          0.38  
6          0.38  
7          0.34  
8           NaN  
9           NaN

In [5]:
adverse_reactions.head(6)

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation
5    jasmine       sykes               hypoglycemia

In [6]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [7]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [8]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [9]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [10]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [11]:
# To see which all rows are null
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [12]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [13]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [14]:
patients.sample(5)

patient_id assigned_sex given_name   surname                     address  \
105         106         male       Nate    Bowden         4310 Johnson Street   
368         369         male      Corey  Nicholls  3427 Gerald L. Bates Drive   
318         319         male     Benoît    Bonami          230 Lady Bug Drive   
65           66         male    Nicolas  Ferreira         2310 Overlook Drive   
28           29         male     Robert      Wolf           2386 Linda Street   

                city     state  zip_code        country  \
105          Clinton        NC   28328.0  United States   
368           Boston        MA    2110.0  United States   
318      Garden City  New York   11530.0  United States   
65         Rossville        IN   46065.0  United States   
28   Fort Washington        PA   19034.0  United States   

                                            contact   birthdate  weight  \
105               NateBowden@einrot.com919-915-8670   3/26/1954   222.0   
368      CoreyNicholls@jourrapide.com1 617 830 7216   4/25/1989   165.0   
318            BenoitBonami@gustr.com1 718 954 8136   4/13/1975   162.8   
65   NicolasSouzaFerreira@superrito.com765-379-3885  10/27/1976   211.9   
28             RobertWolf@fleckens.hu1 267 895 7462   6/26/1937   206.6   

     height   bmi  
105      71  31.0  
368      74  21.2  
318      70  23.4  
65       69  31.3  
28       70  29.6

In [15]:
patients.surname.value_counts()

Doe            6
Jakobsen       3
Taylor         3
Schiavone      2
Gersten        2
Aranda         2
Lương          2
Lund           2
Correia        2
Tạ             2
Grímsdóttir    2
Johnson        2
Dratchev       2
Silva          2
Woźniak        2
Hueber         2
Berg           2
Kowalczyk      2
Tucker         2
Batukayev      2
Cindrić        2
Lâm            2
Liễu           2
Parker         2
Bùi            2
Ogochukwu      2
Collins        2
Souza          2
Kadyrov        2
Cabrera        2
              ..
Sandgreen      1
Mueller        1
Bois           1
Bole           1
Komavec        1
Barsukov       1
Ghanem         1
Crawford       1
Mortensen      1
Christensen    1
Allen          1
Azuma          1
Hughes         1
Hopkins        1
Cabán          1
Pinto          1
Endrizzi       1
Onwughara      1
Lončar         1
Balagić        1
Nebay          1
Vũ             1
Miles          1
Heimisson      1
Fiorentino     1
Ibragimov      1
Þráinsson      1
Isa           

In [16]:
patients.address.value_counts()

123 Main Street                  6
648 Old Dear Lane                2
2476 Fulton Street               2
2778 North Avenue                2
2531 Eastland Avenue             1
4476 Center Street               1
2625 Romines Mill Road           1
4556 Oakdale Avenue              1
1386 Farland Street              1
1257 Elsie Drive                 1
3636 Junior Avenue               1
4324 Oak Street                  1
1051 Bassell Avenue              1
2389 Rubaiyat Road               1
2316 Ingram Road                 1
1067 Ray Court                   1
2172 Lynn Street                 1
1536 John Daniel Drive           1
3595 Stuart Street               1
2704 Windy Ridge Road            1
1624 Rosebud Avenue              1
796 Eagle Street                 1
3548 Hilltop Drive               1
4932 Goldleaf Lane               1
2544 Worley Avenue               1
1953 Rhapsody Street             1
4814 Layman Court                1
383 Marion Street                1
2270 Bel Meadow Driv

In [17]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [18]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [19]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [20]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [21]:
sum(treatments.auralin.isnull())

0

In [22]:
sum(treatments.novodra.isnull())

0

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [23]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data

<font color='red'>Complete the following two "Missing Data" **Define, Code, and Test** sequences after watching the *"Address Missing Data First"* video.</font>

#### `treatments`: Missing records (280 instead of 350)

##### Define
*Your definition here. Note: the missing `treatments` records are stored in a file named `treatments_cut.csv`, which you can see in this Jupyter Notebook's dashboard (click the **jupyter** logo in the top lefthand corner of this Notebook). Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) for the function used in the solution.*

In [24]:
patients.shape

(503, 14)

In [25]:
treatments.shape

(280, 7)

In [26]:
treatments.head(10)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   
5    jasmine       sykes          -  42u - 44u         7.56       7.18   
6     sophia      haugen  37u - 42u          -         7.65       7.27   
7      eddie      archer  31u - 38u          -         7.89       7.55   
8      saber      ménard          -  54u - 54u         8.08       7.70   
9       asia     woźniak  30u - 36u          -         7.76       7.37   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32  
5          0.38  
6          0.38  
7          0.34  
8           NaN  
9           NaN

In [27]:
treatments_cut = pd.read_csv('treatments_cut.csv')
treatments_cut.head(10)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0      jožka   resanovič  22u - 30u          -         7.56       7.22   
1  inunnguaq    heilmann  57u - 67u          -         7.85       7.45   
2      alwin    svensson  36u - 39u          -         7.78       7.34   
3        thể       lương          -  61u - 64u         7.64       7.22   
4     amanda     ribeiro  36u - 44u          -         7.85       7.47   
5       meya     sjöberg          -  42u - 42u         7.77       7.42   
6       sara       miles  30u - 35u          -         7.53       7.12   
7    frydryk     adamski  63u - 74u          -         7.75       7.27   
8       jane     citizen          -  37u - 38u         7.98       7.60   
9     angela  lavrentyev          -  28u - 24u         7.61       7.14   

   hba1c_change  
0          0.34  
1           NaN  
2           NaN  
3          0.92  
4          0.38  
5          0.35  
6           NaN  
7           NaN  
8           NaN  
9          0.97

In [28]:
treatments_cut.shape

(70, 7)

##### Code

In [29]:
# Your cleaning code here
treatments_clean = pd.concat([treatments, treatments_cut], ignore_index=True)
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

##### Test

In [30]:
treatments_clean.shape

(350, 7)

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)
*Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one **Define, Code, and Test** header occurs multiple times in this notebook.*

In [31]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
given_name      350 non-null object
surname         350 non-null object
auralin         350 non-null object
novodra         350 non-null object
hba1c_start     350 non-null float64
hba1c_end       350 non-null float64
hba1c_change    213 non-null float64
dtypes: float64(3), object(4)
memory usage: 19.2+ KB


##### Define
- Subtract 0.5 from Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Compute __hba1c_start - hba1c_end__ for missing hba1c_changes.

##### Code

In [32]:
for i in range(0, len(treatments_clean.hba1c_change)):
    if treatments_clean.hba1c_change[i] >= 0.9:
        treatments_clean.hba1c_change[i] = treatments_clean.hba1c_change[i] - 0.5
        
null_list = treatments_clean.hba1c_change.isnull()
treatments_clean.hba1c_change[null_list] = treatments_clean.hba1c_start[null_list] - treatments_clean.hba1c_end[null_list]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [33]:
treatments_clean.hba1c_change.head(10)

0    0.43
1    0.47
2    0.43
3    0.35
4    0.32
5    0.38
6    0.38
7    0.34
8    0.38
9    0.39
Name: hba1c_change, dtype: float64

##### Test

In [34]:
# Your testing code here
treatments_clean.hba1c_change.sample(10)

169    0.42
240    0.34
46     0.35
309    0.34
208    0.47
37     0.45
203    0.37
279    0.36
308    0.50
18     0.42
Name: hba1c_change, dtype: float64

In [35]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
given_name      350 non-null object
surname         350 non-null object
auralin         350 non-null object
novodra         350 non-null object
hba1c_start     350 non-null float64
hba1c_end       350 non-null float64
hba1c_change    350 non-null float64
dtypes: float64(3), object(4)
memory usage: 19.2+ KB


### Tidiness

<font color='red'>Complete the following four "Tidiness" **Define, Code, and Test** sequences after watching the *"Cleaning for Tidiness"* video.</font>

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
*Your definition here. Hint 1: use regular expressions with pandas' [`str.extract` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html). Here is an amazing [regex tutorial](https://regexone.com/). Hint 2: [various phone number regex patterns](https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number). Hint 3: [email address regex pattern](http://emailregex.com/), which you might need to modify to distinguish the email from the phone number.*

In [36]:
patients['contact'].sample(11)

32                 PhamThiBichMot@dayrep.com562-985-4582
363              562-452-5685ElisabettaLorenzo@gustr.com
83                       937-523-1224ChenYao@teleworm.us
27                   985-814-7603JosephNTucker@rhyta.com
224          802-614-0812VeronikaJindrova@jourrapide.com
11            Abdul-NurMummarIsa@rhyta.com1 931 207 0839
311               EricaMacDonald@teleworm.us601-389-7682
480         NasserMazinMansour@fleckens.hu1 210 326 5509
302                      NielsLange@cuvox.de601-296-7799
408    GiovanaAzevedoRocha@superrito.com+1 (989) 390-...
502                 PatrickGersten@rhyta.com402-848-4923
Name: contact, dtype: object

In [37]:
patients['contact'][415]

'909-861-4356MgbekeEkwueme@cuvox.de'

In [40]:
patients_clean['phone_number'] = patients_clean.contact.str.extract('(\+?\d{1,2}\s)?\(?\d{3}\)?[\s-.]\d{3}[-.\s]\d{4})', expand = True)
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z_]+@[a-z.]+)', expand = True)
# Note: axis=1 denotes that we are referring to a column, not a row
patients_clean = patients_clean.drop('contact', axis=1)

error: bad character range \s-. at position 27

In [39]:
patients['contact'].str.extract('?P<Email>([a-zA-Z_]+@[a-z.]+)(?P<Phone>^')

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)
  """Entry point for launching an IPython kernel.


error: nothing to repeat at position 0

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define
*Your definition here. Hint: use pandas' [melt function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html) and [`str.split()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html). Here is an excellent [`melt` tutorial](https://deparkes.co.uk/2016/10/28/reshape-pandas-data-with-melt/).*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Adverse reaction should be part of the `treatments` table

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for the function used in the solution.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for one function used in the solution and [tutorial](http://erikrood.com/Python_References/dropping_rows_cols_pandas.html) for another function used in the solution.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

### Quality

<font color='red'>Complete the remaining "Quality" **Define, Code, and Test** sequences after watching the *"Cleaning for Quality"* video.</font>

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
*Your definition here. Hint: see the "Data Cleaning Process" page.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Tim Neudorf height is 27 in instead of 72 in

##### Define
*Your definition here.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Full state names sometimes, abbreviations other times

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_apply_operations_to_dataframes.html) for method used in solution.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Dsvid Gustafsson

##### Define
*Your definition here.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define
*Your definition here. Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.astype.html) for one method used in solution, [documentation page](http://pandas.pydata.org/pandas-docs/version/0.20/generated/pandas.to_datetime.html) for one function used in the solution, and [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.strip.html) for another method used in the solution.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Multiple phone number formats

##### Define
*Your definition here. Hint: helpful [Stack Overflow answer](https://stackoverflow.com/a/123681).*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Default John Doe data

##### Define
*Your definition here. Recall that it is assumed that the data that this John Doe data displaced is not recoverable.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
*Your definition here.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### kgs instead of lbs for Zaitseva weight

##### Define
*Your definition here.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here